# Projet Fil Rouge Centrale #

Victor Bonlieu-Fortier, Charles Nikdel

---

## Importation librairies ##

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import utils as ft_nous


In [3]:
rename_columns = [
					"index",
					"date_urg",
					"conduite",
					"m_entry",
					"m_venue",
					"age",
					"atcd_med",
					"atcd_chir",
					"cst_paspad",
					"fc_ad",
					"tmp_ad",
					"saoxy_ad",
					"fr_ad",
					"ccmu",
					"salle",
					"date_box",
					"anamn",
					"obs",
					"have_bio",
					"have_radio",
					"have_echo",
					"have_scan",
					"have_irm",
					"have_specia",
					"date_prescri_bio",
					"date_prelev_bio",
					"code_diag",
					"intitule_diag",
					"cac_admi",
					"cac_sejour",
					"date_sortie",
					"date_deb_pec_med",
					"dossier_hopit",
					"type_orient",
					"destination",
					"transfert_service",
					"transfert_hopit",
					"cp_ville",
					"ville"
					]
rename_columns


['index',
 'date_urg',
 'conduite',
 'm_entry',
 'm_venue',
 'age',
 'atcd_med',
 'atcd_chir',
 'cst_paspad',
 'fc_ad',
 'tmp_ad',
 'saoxy_ad',
 'fr_ad',
 'ccmu',
 'salle',
 'date_box',
 'anamn',
 'obs',
 'have_bio',
 'have_radio',
 'have_echo',
 'have_scan',
 'have_irm',
 'have_specia',
 'date_prescri_bio',
 'date_prelev_bio',
 'code_diag',
 'intitule_diag',
 'cac_admi',
 'cac_sejour',
 'date_sortie',
 'date_deb_pec_med',
 'dossier_hopit',
 'type_orient',
 'destination',
 'transfert_service',
 'transfert_hopit',
 'cp_ville',
 'ville']

In [4]:
df = pd.read_excel("../database/db_urg_liteplus.xlsx", index_col=0, header=0, names = rename_columns)
df


date_urg           conduite  \
index                                          
1     2016-06-01 00:16:00   AMBULANCE PRIVEE   
2     2016-06-01 00:18:00   AMBULANCE PRIVEE   
3     2016-06-01 00:36:00           POMPIERS   
4     2016-06-01 00:43:00  MOYENS PERSONNELS   
5     2016-06-01 01:16:00   AMBULANCE PRIVEE   
...                   ...                ...   
61488 2017-02-26 02:34:00          SAMU-SMUR   
61489 2017-02-26 02:43:00  MOYENS PERSONNELS   
61490 2017-02-26 02:52:00  MOYENS PERSONNELS   
61491 2017-02-26 03:19:00  MOYENS PERSONNELS   
61492 2017-02-26 03:26:00  MOYENS PERSONNELS   

                                                 m_entry  \
index                                                      
1      Malaise, difficulté respi, palpitation, Pas de...   
2      AEG, Hyperthermie et difficulté respi en maiso...   
3                                                    NaN   
4      1A, Hyperthermie ne cédant pas au paracetamol,...   
5      Douleur bras gauche et mollet gauche, hémiplég...   
...                                                  ...   
61488                                                NaN   
61489                                                NaN   
61490                       suspicion cellulite dentaire   
61491                                  intox alimentaire   
61492                  céphalées depuis 1 mois, stat neg   

                                                 m_venue  age  \
index                                                           
1                      Cardiologie: Palpitations Malaise   33   
2                    Maladies infectieuses: Hyperthermie   78   
3                                                    NaN   77   
4                    Maladies infectieuses: Hyperthermie   38   
5                                  Cardiologie: Phlébite   69   
...                                                  ...  ...   
61488                                                NaN   78   
61489                                                NaN   36   
61490            Stomatologie: Abcès/Cellulite dentaire    45   
61491  Gastro-entérologie: Douleur abdominale non épi...   18   
61492                              Neurologie: Céphalées   86   

                                                atcd_med       atcd_chir  \
index                                                                      
1                                                    NaN             NaN   
2                                                    NaN             NaN   
3                                                    NaN             NaN   
4                                                    NaN             NaN   
5      décembre 2007 : infarctus capsulo caudé droit ...             NaN   
...                                                  ...             ...   
61488                                                NaN             NaN   
61489                                                NaN             NaN   
61490            hypertriglycéridémie  crise de goutte    amygdalectomie   
61491                                                NaN             NaN   
61492                           hypertrophie de prostate               0   

      cst_paspad fc_ad tmp_ad  ... cac_sejour   date_sortie date_deb_pec_med  \
index                          ...                                             
1         110/56    52   34.9  ...     3093.0  42522.176389     42522.043750   
2          90/41    77   36.3  ...     3102.0  42522.461806     42522.093056   
3         137/77    97   36.5  ...     3102.0  42522.127778              NaN   
4         117/71   118   38.2  ...     3093.0  42522.201389     42522.081250   
5         137/79    70   36.1  ...     3102.0  42522.663194     42522.107639   
...          ...   ...    ...  ...        ...           ...              ...   
61488      44058    72   36.3  ...     3102.0  42792.137500     42792.108333   
61489        NaN   NaN    NaN  ...     3098.0  42792.122

In [5]:
df.info()


<class 'pandas.core.frame.DataFrame'>
Index: 61492 entries, 1 to 61492
Data columns (total 38 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   date_urg           61492 non-null  datetime64[ns]
 1   conduite           61166 non-null  object        
 2   m_entry            57504 non-null  object        
 3   m_venue            56665 non-null  object        
 4   age                61492 non-null  int64         
 5   atcd_med           16002 non-null  object        
 6   atcd_chir          10989 non-null  object        
 7   cst_paspad         40529 non-null  object        
 8   fc_ad              40680 non-null  object        
 9   tmp_ad             41450 non-null  object        
 10  saoxy_ad           34988 non-null  object        
 11  fr_ad              1692 non-null   object        
 12  ccmu               58861 non-null  object        
 13  salle              61492 non-null  object        
 14  date_box   

On voit que bon, y a beaucoup de colonne mais pas mal qui contienne rien d'interessant, on va donc les supprimer
Dans l'ordre ceux que je pense supprimer et concatenner pour certaines : 

- obs sert pas a grand chose et y en a pas bcq, c'est les observations des inf genre EVA ou autre
- code diag donne le code du diagnostic, on a deja le libelle donc pas besoin
- cac admi jsp vraiment ce que c'est ca doit etre un code de protocole, mais il faut juste récup le numéro comme le cac_sejour
- dossier_hopit y es casi nul donc on peut tej
- type orientation, c'est destination mais moins précis, on peut tej
- transfert service c'est destination mais juste pour les services internes, on peut tej
- transfert etab c'est destination mais juste pour les autres etablissements, y en as pas enorme, je pense on peut tej et et juste modif pour dire autre etab
- le code postal et la ville jsp si c'est utile la vérité à voir, je pense qu'on peut retirer le code postal et garder la ville pour des vizu de zinzin
- frequence respi y en as pas bcq on peut retirer
- Je pense qu'on peut retirer aussi tout ceux qui n'ont pas un code ccmu, ca veut dire ceux qui ont pas été triagé et donc pas pris en charge par les urgences
- salle enfaite c'est une liste de leur voyage dans les urgences, jsp si c'est utile non plus , ca clc à parser et jsp si ca sera utile

Pour réduire le nbr de colonne et le contenu en info :

- Conduite, on peut changer ca en 1 2 3 4 5 ... y 'en a pas bcq
- motif entrée et admission : clean et tokenizer comme on as fait sur twitter
- atcd med et chir idem 
- les constantes jsp pq aucune est en int il faut convert : trouver un truc pck la pression sanguine en 110 / 56 ca clc par exemple vu que c'est pas un int
- convertir la température en int, pour remplacer les NaN juste mettre la température normale
- la saturation en oxy, pareil remplacer les NaN par la saturation normale et convertir en int sachant que y a des inf clc qui ont met AA (air ambiant) du coup les retirer
- anamnèse, on peut tokenizer et clean comme on as fait sur twitter
- tout les have, à transformer en 0 1
- les date prescri bio et prelev, peut être juste faire un delta entre les deux, y en as ils sont pas prescri donc on peut mettre 0 et d'autre qui sont prescri mais pas prelevé donc mettre la date de prescri ou la moyenne des dates de prescri jpense
- intitule diag à tokenizer et clean
- parser destination pour juste garder le service / hopital / autre hopital
- ville ca a l'air d'être bon

In [6]:
df.columns


Index(['date_urg', 'conduite', 'm_entry', 'm_venue', 'age', 'atcd_med',
       'atcd_chir', 'cst_paspad', 'fc_ad', 'tmp_ad', 'saoxy_ad', 'fr_ad',
       'ccmu', 'salle', 'date_box', 'anamn', 'obs', 'have_bio', 'have_radio',
       'have_echo', 'have_scan', 'have_irm', 'have_specia', 'date_prescri_bio',
       'date_prelev_bio', 'code_diag', 'intitule_diag', 'cac_admi',
       'cac_sejour', 'date_sortie', 'date_deb_pec_med', 'dossier_hopit',
       'type_orient', 'destination', 'transfert_service', 'transfert_hopit',
       'cp_ville', 'ville'],
      dtype='object')

In [7]:
df = df.dropna(subset=["ccmu","m_venue"])
df = df.drop(["obs","anamn","salle","m_entry","atcd_chir", "atcd_med","code_diag","dossier_hopit","type_orient","transfert_service","transfert_hopit","fr_ad","ville"],axis=1)
df.info()


<class 'pandas.core.frame.DataFrame'>
Index: 55698 entries, 1 to 61492
Data columns (total 25 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   date_urg          55698 non-null  datetime64[ns]
 1   conduite          55683 non-null  object        
 2   m_venue           55698 non-null  object        
 3   age               55698 non-null  int64         
 4   cst_paspad        38311 non-null  object        
 5   fc_ad             38467 non-null  object        
 6   tmp_ad            39290 non-null  object        
 7   saoxy_ad          32996 non-null  object        
 8   ccmu              55698 non-null  object        
 9   date_box          32568 non-null  float64       
 10  have_bio          55698 non-null  object        
 11  have_radio        55698 non-null  object        
 12  have_echo         55698 non-null  object        
 13  have_scan         55698 non-null  object        
 14  have_irm          55698 non

In [8]:
df = ft_nous.lstdate(df)
df = ft_nous.convertir_conduite(df,"conduite")
df = ft_nous.pipelinett(df)
df.head()

date_urg  conduite                              m_venue  age  \
index                                                                           
1     2016-06-01 00:16:00         2    Cardiologie: Palpitations Malaise   33   
2     2016-06-01 00:18:00         2  Maladies infectieuses: Hyperthermie   78   
4     2016-06-01 00:43:00         6  Maladies infectieuses: Hyperthermie   38   
5     2016-06-01 01:16:00         2                Cardiologie: Phlébite   69   
6     2016-06-01 01:20:00         2                        Orl: Vertiges   80   

      fc_ad  tmp_ad  saoxy_ad ccmu                   date_box have_bio  ...  \
index                                                                   ...   
1        52    34.9        97    1 2016-06-01 01:05:59.999997      oui  ...   
2        77    36.3        98    2 2016-06-01 01:32:00.000001      oui  ...   
4       118    38.2        98    1 2016-06-01 01:36:00.000003      oui  ...   
5        70    36.1        97    2 2016-06-01 01:36:00.000003      oui  ...   
6        97    36.4        97    1 2016-06-01 01:54:59.999999      non  ...   

                 date_prelev_bio  \
index                              
1     2016-06-01 01:34:59.999997   
2     2016-06-01 01:34:00.000002   
4     2016-06-01 01:39:59.999996   
5     2016-06-01 01:34:59.999997   
6     1899-12-30 00:00:00.000000   

                                           intitule_diag cac_admi cac_sejour  \
index                                                                          
1                                                Dyspnée     3093     3093.0   
2                                 Fièvre, sans précision     3093     3102.0   
4      Autres affections précisées de la peau et du t...     3093     3093.0   
5                   Douleur au niveau d'un membre - Bras     3093     3102.0   
6                                             Acouphènes     3093     3093.0   

                     date_sortie           date_deb_pec_med destination  \
index                                                                     
1     2016-06-01 04:14:00.000001 2016-06-01 01:02:59.999999           1   
2     2016-06-01 11:05:00.000003 2016-06-01 02:13:59.999995           4   
4     2016-06-01 04:50:00.000000 2016-06-01 01:57:00.000000           1   
5     2016-06-01 15:54:59.999996 2016-06-01 02:35:00.000000           1   
6     2016-06-01 04:36:59.999998 2016-06-01 03:00:00.000000           1   

      cp_ville  systo  diasto  
index                          
1      59139.0  110.0    56.0  
2      59310.0   90.0    41.0  
4      59136.0  117.0    71.0  
5      59130.0  137.0    79.0  
6      59320.0  149.0    89.0  

[5 rows x 26 columns]

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 55698 entries, 1 to 61492
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   date_urg          55698 non-null  datetime64[ns]
 1   conduite          55698 non-null  int64         
 2   m_venue           55698 non-null  object        
 3   age               55698 non-null  int64         
 4   fc_ad             38467 non-null  object        
 5   tmp_ad            39290 non-null  float64       
 6   saoxy_ad          55698 non-null  int64         
 7   ccmu              55698 non-null  object        
 8   date_box          55698 non-null  datetime64[ns]
 9   have_bio          55698 non-null  object        
 10  have_radio        55698 non-null  object        
 11  have_echo         55698 non-null  object        
 12  have_scan         55698 non-null  object        
 13  have_irm          55698 non-null  object        
 14  have_specia       55698 non

Maintenant je vais faire des fonctions pour traiter les données et les remettre dans le bon type en général

In [10]:
df.head()

date_urg  conduite                              m_venue  age  \
index                                                                           
1     2016-06-01 00:16:00         2    Cardiologie: Palpitations Malaise   33   
2     2016-06-01 00:18:00         2  Maladies infectieuses: Hyperthermie   78   
4     2016-06-01 00:43:00         6  Maladies infectieuses: Hyperthermie   38   
5     2016-06-01 01:16:00         2                Cardiologie: Phlébite   69   
6     2016-06-01 01:20:00         2                        Orl: Vertiges   80   

      fc_ad  tmp_ad  saoxy_ad ccmu                   date_box have_bio  ...  \
index                                                                   ...   
1        52    34.9        97    1 2016-06-01 01:05:59.999997      oui  ...   
2        77    36.3        98    2 2016-06-01 01:32:00.000001      oui  ...   
4       118    38.2        98    1 2016-06-01 01:36:00.000003      oui  ...   
5        70    36.1        97    2 2016-06-01 01:36:00.000003      oui  ...   
6        97    36.4        97    1 2016-06-01 01:54:59.999999      non  ...   

                 date_prelev_bio  \
index                              
1     2016-06-01 01:34:59.999997   
2     2016-06-01 01:34:00.000002   
4     2016-06-01 01:39:59.999996   
5     2016-06-01 01:34:59.999997   
6     1899-12-30 00:00:00.000000   

                                           intitule_diag cac_admi cac_sejour  \
index                                                                          
1                                                Dyspnée     3093     3093.0   
2                                 Fièvre, sans précision     3093     3102.0   
4      Autres affections précisées de la peau et du t...     3093     3093.0   
5                   Douleur au niveau d'un membre - Bras     3093     3102.0   
6                                             Acouphènes     3093     3093.0   

                     date_sortie           date_deb_pec_med destination  \
index                                                                     
1     2016-06-01 04:14:00.000001 2016-06-01 01:02:59.999999           1   
2     2016-06-01 11:05:00.000003 2016-06-01 02:13:59.999995           4   
4     2016-06-01 04:50:00.000000 2016-06-01 01:57:00.000000           1   
5     2016-06-01 15:54:59.999996 2016-06-01 02:35:00.000000           1   
6     2016-06-01 04:36:59.999998 2016-06-01 03:00:00.000000           1   

      cp_ville  systo  diasto  
index                          
1      59139.0  110.0    56.0  
2      59310.0   90.0    41.0  
4      59136.0  117.0    71.0  
5      59130.0  137.0    79.0  
6      59320.0  149.0    89.0  

[5 rows x 26 columns]

In [11]:
from sklearn.preprocessing import LabelEncoder

l_e = LabelEncoder()
df.head()

date_urg  conduite                              m_venue  age  \
index                                                                           
1     2016-06-01 00:16:00         2    Cardiologie: Palpitations Malaise   33   
2     2016-06-01 00:18:00         2  Maladies infectieuses: Hyperthermie   78   
4     2016-06-01 00:43:00         6  Maladies infectieuses: Hyperthermie   38   
5     2016-06-01 01:16:00         2                Cardiologie: Phlébite   69   
6     2016-06-01 01:20:00         2                        Orl: Vertiges   80   

      fc_ad  tmp_ad  saoxy_ad ccmu                   date_box have_bio  ...  \
index                                                                   ...   
1        52    34.9        97    1 2016-06-01 01:05:59.999997      oui  ...   
2        77    36.3        98    2 2016-06-01 01:32:00.000001      oui  ...   
4       118    38.2        98    1 2016-06-01 01:36:00.000003      oui  ...   
5        70    36.1        97    2 2016-06-01 01:36:00.000003      oui  ...   
6        97    36.4        97    1 2016-06-01 01:54:59.999999      non  ...   

                 date_prelev_bio  \
index                              
1     2016-06-01 01:34:59.999997   
2     2016-06-01 01:34:00.000002   
4     2016-06-01 01:39:59.999996   
5     2016-06-01 01:34:59.999997   
6     1899-12-30 00:00:00.000000   

                                           intitule_diag cac_admi cac_sejour  \
index                                                                          
1                                                Dyspnée     3093     3093.0   
2                                 Fièvre, sans précision     3093     3102.0   
4      Autres affections précisées de la peau et du t...     3093     3093.0   
5                   Douleur au niveau d'un membre - Bras     3093     3102.0   
6                                             Acouphènes     3093     3093.0   

                     date_sortie           date_deb_pec_med destination  \
index                                                                     
1     2016-06-01 04:14:00.000001 2016-06-01 01:02:59.999999           1   
2     2016-06-01 11:05:00.000003 2016-06-01 02:13:59.999995           4   
4     2016-06-01 04:50:00.000000 2016-06-01 01:57:00.000000           1   
5     2016-06-01 15:54:59.999996 2016-06-01 02:35:00.000000           1   
6     2016-06-01 04:36:59.999998 2016-06-01 03:00:00.000000           1   

      cp_ville  systo  diasto  
index                          
1      59139.0  110.0    56.0  
2      59310.0   90.0    41.0  
4      59136.0  117.0    71.0  
5      59130.0  137.0    79.0  
6      59320.0  149.0    89.0  

[5 rows x 26 columns]

In [12]:
df["m_venue"] = l_e.fit_transform(df["m_venue"])
df.head()

date_urg  conduite  m_venue  age fc_ad  tmp_ad  saoxy_ad  \
index                                                                       
1     2016-06-01 00:16:00         2      391   33    52    34.9        97   
2     2016-06-01 00:18:00         2     1261   78    77    36.3        98   
4     2016-06-01 00:43:00         6     1261   38   118    38.2        98   
5     2016-06-01 01:16:00         2      421   69    70    36.1        97   
6     2016-06-01 01:20:00         2     1621   80    97    36.4        97   

      ccmu                   date_box have_bio  ...  \
index                                           ...   
1        1 2016-06-01 01:05:59.999997      oui  ...   
2        2 2016-06-01 01:32:00.000001      oui  ...   
4        1 2016-06-01 01:36:00.000003      oui  ...   
5        2 2016-06-01 01:36:00.000003      oui  ...   
6        1 2016-06-01 01:54:59.999999      non  ...   

                 date_prelev_bio  \
index                              
1     2016-06-01 01:34:59.999997   
2     2016-06-01 01:34:00.000002   
4     2016-06-01 01:39:59.999996   
5     2016-06-01 01:34:59.999997   
6     1899-12-30 00:00:00.000000   

                                           intitule_diag cac_admi cac_sejour  \
index                                                                          
1                                                Dyspnée     3093     3093.0   
2                                 Fièvre, sans précision     3093     3102.0   
4      Autres affections précisées de la peau et du t...     3093     3093.0   
5                   Douleur au niveau d'un membre - Bras     3093     3102.0   
6                                             Acouphènes     3093     3093.0   

                     date_sortie           date_deb_pec_med destination  \
index                                                                     
1     2016-06-01 04:14:00.000001 2016-06-01 01:02:59.999999           1   
2     2016-06-01 11:05:00.000003 2016-06-01 02:13:59.999995           4   
4     2016-06-01 04:50:00.000000 2016-06-01 01:57:00.000000           1   
5     2016-06-01 15:54:59.999996 2016-06-01 02:35:00.000000           1   
6     2016-06-01 04:36:59.999998 2016-06-01 03:00:00.000000           1   

      cp_ville  systo  diasto  
index                          
1      59139.0  110.0    56.0  
2      59310.0   90.0    41.0  
4      59136.0  117.0    71.0  
5      59130.0  137.0    79.0  
6      59320.0  149.0    89.0  

[5 rows x 26 columns]

In [13]:
df["intitule_diag"] = l_e.fit_transform(df["intitule_diag"])
df.head()

date_urg  conduite  m_venue  age fc_ad  tmp_ad  saoxy_ad  \
index                                                                       
1     2016-06-01 00:16:00         2      391   33    52    34.9        97   
2     2016-06-01 00:18:00         2     1261   78    77    36.3        98   
4     2016-06-01 00:43:00         6     1261   38   118    38.2        98   
5     2016-06-01 01:16:00         2      421   69    70    36.1        97   
6     2016-06-01 01:20:00         2     1621   80    97    36.4        97   

      ccmu                   date_box have_bio  ...  \
index                                           ...   
1        1 2016-06-01 01:05:59.999997      oui  ...   
2        2 2016-06-01 01:32:00.000001      oui  ...   
4        1 2016-06-01 01:36:00.000003      oui  ...   
5        2 2016-06-01 01:36:00.000003      oui  ...   
6        1 2016-06-01 01:54:59.999999      non  ...   

                 date_prelev_bio intitule_diag cac_admi cac_sejour  \
index                                                                
1     2016-06-01 01:34:59.999997          1187     3093     3093.0   
2     2016-06-01 01:34:00.000002          1365     3093     3102.0   
4     2016-06-01 01:39:59.999996           377     3093     3093.0   
5     2016-06-01 01:34:59.999997          1152     3093     3102.0   
6     1899-12-30 00:00:00.000000            61     3093     3093.0   

                     date_sortie           date_deb_pec_med destination  \
index                                                                     
1     2016-06-01 04:14:00.000001 2016-06-01 01:02:59.999999           1   
2     2016-06-01 11:05:00.000003 2016-06-01 02:13:59.999995           4   
4     2016-06-01 04:50:00.000000 2016-06-01 01:57:00.000000           1   
5     2016-06-01 15:54:59.999996 2016-06-01 02:35:00.000000           1   
6     2016-06-01 04:36:59.999998 2016-06-01 03:00:00.000000           1   

       cp_ville  systo  diasto  
index                           
1       59139.0  110.0    56.0  
2       59310.0   90.0    41.0  
4       59136.0  117.0    71.0  
5       59130.0  137.0    79.0  
6       59320.0  149.0    89.0  

[5 rows x 26 columns]

In [14]:
df.columns

Index(['date_urg', 'conduite', 'm_venue', 'age', 'fc_ad', 'tmp_ad', 'saoxy_ad',
       'ccmu', 'date_box', 'have_bio', 'have_radio', 'have_echo', 'have_scan',
       'have_irm', 'have_specia', 'date_prescri_bio', 'date_prelev_bio',
       'intitule_diag', 'cac_admi', 'cac_sejour', 'date_sortie',
       'date_deb_pec_med', 'destination', 'cp_ville', 'systo', 'diasto'],
      dtype='object')

In [15]:
print(df["destination"].value_counts())

destination
1       42821
4        1678
2250      637
2         609
3715      468
        ...  
4260        1
3344        1
3839        1
3959        1
4245        1
Name: count, Length: 289, dtype: int64


In [16]:
tlt = ["date_urg","date_box","date_prescri_bio","date_prelev_bio","date_sortie","date_deb_pec_med"]
for i in tlt:
    df[i] = pd.to_datetime(df[i], format='%Y-%m-%d %H:%M:%S')

def minutes_since_midnight(time):
    return time.hour * 60 + time.minute

# Fonction pour calculer le delta en minutes
def calculate_delta(start, end):
    
    delta = pd.to_datetime(end) - pd.to_datetime(start)
    if delta.total_seconds() <= 0:
        return np.nan
    return delta.total_seconds() / 60

# Calculs
df['arrivee_minutes'] = df['date_urg'].apply(minutes_since_midnight)
df['delta_arrivee_pec_med'] = df.apply(lambda row: calculate_delta(row['date_urg'], row['date_deb_pec_med']), axis=1)
df['delta_arrivee_box'] = df.apply(lambda row: calculate_delta(row['date_urg'], row['date_box']), axis=1)
df['delta_prescri_bio'] = df.apply(lambda row: calculate_delta(row['date_urg'], row['date_prescri_bio']), axis=1)
df['duree_prelev_bio'] = df.apply(lambda row: calculate_delta(row['date_prescri_bio'], row['date_prelev_bio']), axis=1)
df['delta_arrivee_sortie'] = df.apply(lambda row: calculate_delta(row['date_urg'], row['date_sortie']), axis=1)

df.head()


date_urg  conduite  m_venue  age fc_ad  tmp_ad  saoxy_ad  \
index                                                                       
1     2016-06-01 00:16:00         2      391   33    52    34.9        97   
2     2016-06-01 00:18:00         2     1261   78    77    36.3        98   
4     2016-06-01 00:43:00         6     1261   38   118    38.2        98   
5     2016-06-01 01:16:00         2      421   69    70    36.1        97   
6     2016-06-01 01:20:00         2     1621   80    97    36.4        97   

      ccmu                   date_box have_bio  ... destination cp_ville  \
index                                           ...                        
1        1 2016-06-01 01:05:59.999997      oui  ...           1  59139.0   
2        2 2016-06-01 01:32:00.000001      oui  ...           4  59310.0   
4        1 2016-06-01 01:36:00.000003      oui  ...           1  59136.0   
5        2 2016-06-01 01:36:00.000003      oui  ...           1  59130.0   
6        1 2016-06-01 01:54:59.999999      non  ...           1  59320.0   

       systo diasto arrivee_minutes delta_arrivee_pec_med delta_arrivee_box  \
index                                                                         
1      110.0   56.0              16                  47.0              50.0   
2       90.0   41.0              18                 116.0              74.0   
4      117.0   71.0              43                  74.0              53.0   
5      137.0   79.0              76                  79.0              20.0   
6      149.0   89.0              80                 100.0              35.0   

       delta_prescri_bio duree_prelev_bio  delta_arrivee_sortie  
index                                                            
1                   74.0              5.0                 238.0  
2                   76.0              NaN                 647.0  
4                   57.0              NaN                 247.0  
5                   18.0              1.0                 879.0  
6                    NaN              NaN                 197.0  

[5 rows x 32 columns]

In [17]:
txt = ['have_bio', 'have_radio', 'have_echo', 'have_scan','have_irm', 'have_specia']
for x in txt:
	df[x] = l_e.fit_transform(df[x])
df.head()

date_urg  conduite  m_venue  age fc_ad  tmp_ad  saoxy_ad  \
index                                                                       
1     2016-06-01 00:16:00         2      391   33    52    34.9        97   
2     2016-06-01 00:18:00         2     1261   78    77    36.3        98   
4     2016-06-01 00:43:00         6     1261   38   118    38.2        98   
5     2016-06-01 01:16:00         2      421   69    70    36.1        97   
6     2016-06-01 01:20:00         2     1621   80    97    36.4        97   

      ccmu                   date_box  have_bio  ...  destination  cp_ville  \
index                                            ...                          
1        1 2016-06-01 01:05:59.999997         1  ...            1   59139.0   
2        2 2016-06-01 01:32:00.000001         1  ...            4   59310.0   
4        1 2016-06-01 01:36:00.000003         1  ...            1   59136.0   
5        2 2016-06-01 01:36:00.000003         1  ...            1   59130.0   
6        1 2016-06-01 01:54:59.999999         0  ...            1   59320.0   

       systo  diasto  arrivee_minutes delta_arrivee_pec_med delta_arrivee_box  \
index                                                                           
1      110.0    56.0               16                  47.0              50.0   
2       90.0    41.0               18                 116.0              74.0   
4      117.0    71.0               43                  74.0              53.0   
5      137.0    79.0               76                  79.0              20.0   
6      149.0    89.0               80                 100.0              35.0   

       delta_prescri_bio duree_prelev_bio  delta_arrivee_sortie  
index                                                            
1                   74.0              5.0                 238.0  
2                   76.0              NaN                 647.0  
4                   57.0              NaN                 247.0  
5                   18.0              1.0                 879.0  
6                    NaN              NaN                 197.0  

[5 rows x 32 columns]

In [18]:
df = df.drop(["date_urg","date_box","date_prescri_bio","date_prelev_bio","date_sortie","date_deb_pec_med"],axis=1)

In [19]:
def round_up_to_nearest_15(minutes):
    return np.ceil(minutes / 15) * 15

x = ["arrivee_minutes","delta_arrivee_pec_med","delta_arrivee_box","delta_prescri_bio","duree_prelev_bio","delta_arrivee_sortie"]
for t in x:
    df[t] = df[t].apply(round_up_to_nearest_15)
df.head()

conduite  m_venue  age fc_ad  tmp_ad  saoxy_ad ccmu  have_bio  \
index                                                                  
1             2      391   33    52    34.9        97    1         1   
2             2     1261   78    77    36.3        98    2         1   
4             6     1261   38   118    38.2        98    1         1   
5             2      421   69    70    36.1        97    2         1   
6             2     1621   80    97    36.4        97    1         0   

       have_radio  have_echo  ...  destination  cp_ville  systo  diasto  \
index                         ...                                         
1               1          0  ...            1   59139.0  110.0    56.0   
2               1          0  ...            4   59310.0   90.0    41.0   
4               0          0  ...            1   59136.0  117.0    71.0   
5               0          0  ...            1   59130.0  137.0    79.0   
6               0          0  ...            1   59320.0  149.0    89.0   

      arrivee_minutes  delta_arrivee_pec_med delta_arrivee_box  \
index                                                            
1                30.0                   60.0              60.0   
2                30.0                  120.0              75.0   
4                45.0                   75.0              60.0   
5                90.0                   90.0              30.0   
6                90.0                  105.0              45.0   

       delta_prescri_bio  duree_prelev_bio  delta_arrivee_sortie  
index                                                             
1                   75.0              15.0                 240.0  
2                   90.0               NaN                 660.0  
4                   60.0               NaN                 255.0  
5                   30.0              15.0                 885.0  
6                    NaN               NaN                 210.0  

[5 rows x 26 columns]

In [20]:
df['fc_ad'] = df['fc_ad'].apply(lambda x: np.nan if not isinstance(x, int) else x)
#df = df.drop(["cd_ad"],axis=1)

In [21]:
df["ccmu"] = df["ccmu"].apply(lambda x: np.nan if not isinstance(x, int) else x)

In [22]:
from sklearn.impute import SimpleImputer

x = ["fc_ad","tmp_ad","systo","diasto","delta_arrivee_pec_med","delta_arrivee_box","delta_prescri_bio","duree_prelev_bio"]
imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
df = pd.DataFrame(imputer.fit_transform(df), columns=df.columns)
df.head()

conduite  m_venue   age  fc_ad  tmp_ad  saoxy_ad  ccmu  have_bio  \
0       2.0    391.0  33.0   52.0    34.9      97.0   1.0       1.0   
1       2.0   1261.0  78.0   77.0    36.3      98.0   2.0       1.0   
2       6.0   1261.0  38.0  118.0    38.2      98.0   1.0       1.0   
3       2.0    421.0  69.0   70.0    36.1      97.0   2.0       1.0   
4       2.0   1621.0  80.0   97.0    36.4      97.0   1.0       0.0   

   have_radio  have_echo  ...  destination  cp_ville  systo  diasto  \
0         1.0        0.0  ...          1.0   59139.0  110.0    56.0   
1         1.0        0.0  ...          4.0   59310.0   90.0    41.0   
2         0.0        0.0  ...          1.0   59136.0  117.0    71.0   
3         0.0        0.0  ...          1.0   59130.0  137.0    79.0   
4         0.0        0.0  ...          1.0   59320.0  149.0    89.0   

   arrivee_minutes  delta_arrivee_pec_med  delta_arrivee_box  \
0             30.0                   60.0               60.0   
1             30.0                  120.0               75.0   
2             45.0                   75.0               60.0   
3             90.0                   90.0               30.0   
4             90.0                  105.0               45.0   

   delta_prescri_bio  duree_prelev_bio  delta_arrivee_sortie  
0          75.000000          15.00000                 240.0  
1          90.000000          31.12444                 660.0  
2          60.000000          31.12444                 255.0  
3          30.000000          15.00000                 885.0  
4         104.795392          31.12444                 210.0  

[5 rows x 26 columns]

In [23]:
from sklearn.preprocessing import StandardScaler

x = df.drop("delta_arrivee_sortie", axis = 1)
y = df["delta_arrivee_sortie"]
# Standardisation des caractéristiques (excluant la cible)
X_scaled = StandardScaler().fit_transform(x)

In [24]:
from sklearn.decomposition import PCA

# Choix du nombre de composantes
pca = PCA(n_components=5)  # Exemple : réduire à 10 composantes principales


X_pca = pca.fit_transform(X_scaled)
print(pca.explained_variance_ratio_)

# Variance cumulée
cumulative_variance = np.cumsum(pca.explained_variance_ratio_)
print(cumulative_variance)

[0.12869706 0.06613132 0.05700028 0.0518701  0.04827707]
[0.12869706 0.19482837 0.25182865 0.30369875 0.35197582]


In [25]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_pca, y, test_size=0.2, random_state=42)

In [26]:
from sklearn.svm import SVR as SVM

model = SVM()
model.fit(X_train, y_train)

SVR()

In [27]:
from sklearn.metrics import mean_squared_error, r2_score

# Prédiction sur l'ensemble de test
y_pred = model.predict(X_test)

# Calcul des métriques
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"MSE: {mse}")
print(f"R²: {r2}")

MSE: 713327.0095044753
R²: 0.07539944054342573


In [28]:
# y_pred représente les prédictions de votre modèle
# y_true représente les valeurs réelles

diff = np.abs(y_pred - y_test)
tolerance = 15  # 15 minutes
within_tolerance = np.sum(diff <= tolerance)

percentage_within_tolerance = (within_tolerance / len(y_test)) * 100
print(f"Pourcentage de prédictions dans un seuil de tolérance de ±{tolerance} minutes: {percentage_within_tolerance:.2f}%")



Pourcentage de prédictions dans un seuil de tolérance de ±15 minutes: 12.44%
